[Update]: Commit new error message API and needed refactoring

KTSnowy opened the pull request:
This PR contains the new error message API for Otterkit, as well as some refactoring needed to support the new API. This provides a modern look for Otterkit's error messages, which should massively improve the developer experience around debugging and fixing code issues, and also provide useful, easy to read information for the users.

Example of how the messages currently looks like:
![Screenshot 2023-03-24 112903](https://user-images.githubusercontent.com/73310001/227563262-b0405912-ea81-4264-91aa-48b85e117c02.png)


The current API is based around method chaining on a ref struct, meaning it shouldn't allocate any new objects on the heap (aside from the strings) and should perform reasonably well:
![Screenshot 2023-03-24 120611](https://user-images.githubusercontent.com/73310001/227564508-affd76f7-cd25-48eb-80d9-a83f48252858.png)

The API allows for the messages to be later extended with additional optional information. The above `.WithNote` method for example, is optional. Without the method, the error message would look like this:
![Screenshot 2023-03-24 113240](https://user-images.githubusercontent.com/73310001/227566502-2b4dcb2d-9065-4d04-9222-091961e776fb.png)

This should provide a great way to display useful information to experienced developers, while still being easy to read and understand for new developers.

KTSnowy commented:
Leaving another example here for future reference. With some extra fine tuning Otterkit can display some great error messages back.
![Screenshot 2023-03-24 201459](https://user-images.githubusercontent.com/73310001/227667132-cef7677b-e608-4c82-8d7d-f2d1bb5f978b.png)

GitMensch commented:
As you've asked: that looks nice.

Did you already implemented a switch to return those in JSON and/or XML to improve automated parsing (IDE/CI)?

How would the source reference look like if the code is the result of a COPY REPLACING from a nested copybook?

How would notes with source references look like? Your last example may want to show the original PROGRAM-ID line, for example.

GitMensch commented:
For the error display you possibly want an alternative to specify a range of two tokens and possibly also a manual range.
The latter would allow to show the missing period on the right place (after the token), otherwise you could create an error token which is one character wide, has a space as text and is inserted directly after the DIVISION token.

KTSnowy commented:
Hey @GitMensch.
> Did you already implemented a switch to return those in JSON and/or XML to improve automated parsing (IDE/CI)?

Not yet, still need to figure out which JSON format the compiler should return, and also how to deal with the optional parts of the new error messages.

> How would the source reference look like if the code is the result of a COPY REPLACING from a nested copybook?

Currently it doesn't work, the compiler can't yet fetch the copybook name for the errors so it displays back the wrong source line and source file.

I'm not sure how to deal with copybooks with the source reference. Should we reference both the line where the `COPY` is defined and the line inside the copybooks?

> How would notes with source references look like?

Haven't implemented notes with source references yet, but I'll work on it today. I'll post an example of it here later today.

KTSnowy commented:
> For the error display you possibly want an alternative to specify a range of two tokens and possibly also a manual range.

That's a great idea, I'll work on adding those as overloads to the `WithSourceLine` method. I'll post examples of these here as well.

For the inserted token, I think I have an idea of how to implement that. The inserted token should probably be colored red, right? That way it won't be confused with an existing one.

KTSnowy commented:
Hey @GitMensch, here's an example of notes with source references:
![Screenshot 2023-03-25 192118](https://user-images.githubusercontent.com/73310001/227745371-86b6ee99-fc5e-491f-9d83-5b756b599fa8.png)

Had to do a little refactoring to keep track of the exact token used for the program definition. It seems to work nicely.

KTSnowy commented:
The API is still quite simple to build an error with notes to a source line.
![Screenshot 2023-03-25 193348](https://user-images.githubusercontent.com/73310001/227745494-f6a0b508-a8dc-4ee6-b463-d084103ca133.png)

We can keep extending it like this to create some amazing diagnostic messages.

GitMensch commented:
> Possible fix for this could be adding another property to the tokens that reference which file it belongs to. That way it won't matter how nested it is in a copybook, the compiler will always know where the token came from

If it doesn't take too much computing power to track this for each token, then this sound like a good idea - because then you can even drop the `Current (), Filename` from `.WithSourceLine` (which would only override the token reference, if still used).

KTSnowy commented:
> If it doesn't take too much computing power to track this for each token

This should be fine as long as we can make sure that each token will hold a reference to the same string in memory for each file path instead of creating new strings for each token.

The compiler already keeps track of the file names, so the new property could just be a reference to a file in the list of paths we currently have.

KTSnowy commented:
Hey @GitMensch, just implemented the token-based file name tracking.

Added a new field to the Token type that holds an integer index:
![Screenshot 2023-03-26 215333](https://user-images.githubusercontent.com/73310001/227816714-11746b7f-cd21-48c1-9c7f-289091b96e6b.png)

The index is used in a `.FetchFile()` method for the Token type:
![Screenshot 2023-03-26 215355](https://user-images.githubusercontent.com/73310001/227816773-04929af2-dc12-4882-b6dc-dd1619a01387.png)

The method just fetches the file name from the already existing List of file names that the compiler uses for the error messages. 
This means that there should be almost no extra memory or computing overhead for this (aside from the extra integer and method call):
![Screenshot 2023-03-26 215603](https://user-images.githubusercontent.com/73310001/227816890-7b5a686d-77e1-4e87-ab6b-804e1f870d22.png)

The index is assigned to each token during the preprocessing and tokenization of each file, this means that each token should know exactly where it came from, no matter how nested the copybooks are.
